src/v/kafka/server/handlers/fetch.cc

```diff
@@ -143,6 +144,101 @@ static ss::future<read_result> read_from_partition(
       std::move(aborted_transactions));
 }
 
+/**
+ * Consume proper amounts of units from memory semaphores and return them as
+ * semaphore_units. Fetch semaphore units returned are the indication of
+ * available resources: if none, there is no memory for the operation;
+ * if less than \p max_bytes, the fetch should be capped to that size.
+ *
+ * \param max_bytes The limit of how much data is going to be fetched
+ * \param obligatory_batch_read Set to true for the first ntp in the fetch
+ *   fetch request, at least one batch must be fetched for that ntp. Also it
+ *   is assumed that a batch size has already been consumed from kafka
```

Comment:
>Set to true for the first ntp in the fetch fetch request, at least one batch must be fetched for that ntp.

does a similar scenario happen after the first ntp? that is, if we end up blocking part of a subsequent request because we ran out of semaphore units, then, could it be the case that even if all the semaphore units were returned that the request was large enough that it exceeded the units? i'm thinking like: this isn't a bucket that eventually fills up, so we may need to always be on the look out for letting a request through?